bitbybit/src/bitenum.rs

```diff
@@ -1,303 +1,225 @@
-use proc_macro::TokenStream;
-use std::str::FromStr;
-
-use proc_macro2::TokenTree;
-use quote::{quote, ToTokens};
-use syn::__private::TokenStream2;
-use syn::{Attribute, Data, DeriveInput, Expr, Ident};
-
-pub fn bitenum(args: TokenStream, input: TokenStream) -> TokenStream {
-    let args: Vec<_> = proc_macro2::TokenStream::from(args).into_iter().collect();
-
-    let mut bits: Option<usize> = None;
-    let mut exhaustive_value: Option<TokenStream2> = None;
+use proc_macro2::{Span, TokenStream};
+use quote::{quote, spanned::Spanned, ToTokens};
+use syn::{meta::ParseNestedMeta, Token};
+
+use crate::bit_size::Bits;
+
+const BAD_EXHAUSTIVE: &'static str = "The specified exhaustiveness is invalid, supported values are 'true' 'false' and 'conditional'";
+const WRONG_EXHAUSTIVE: &'static str = "The specified 'exhaustive' is invalid";
+const MISSING_EXHAUSTIVE: &'static str = "Missing the 'exhaustive = …' argument to bitenum";
+const MISSING_SIZE: &'static str = "Missing the 'u[0-9]' argument to bitenum";
+const MISSING_DISCR: &'static str =
+    "ALL variants of a #[bitenum] MUST have an explicit literal discriminant eg: 'Variant = 1'.";
+const UNEXPECTED_ATTRIBUTE_META: &'static str = "Invalid bitenum attribute, expected either `uN` where N in range 1..=64, or `exhaustive = (true|false|conditional)`";
+const CFG_NON_CONDITIONAL: &'static str =
+    "The specified exhaustiveness is invalid, the enum contains at least one \
+    conditional variant, therefore, exhaustiveness should be specified as 'conditional'";
+const NONLIT_DISCR: &'static str =
+    "Discriminants must be literals integers, either binary, hexadecimal, octal or decimal";
+const TOO_MANY_VARIANTS: &'static str =
+    "The enum has more variants than can be stored within the provided storage type, consider \
+    using a larger storage type or reducing the number of variants.";
+const DISCR_TOO_LARGE: &'static str =
+    "The largest discriminant value is larger than can be stored in the provided storage type.";
+
+#[derive(Clone, Copy)]
+enum Exhaustiveness {
+    True,
+    False,
+    Conditional,
+}
+struct Exhaustive {
+    span: Span,
+    kind: Exhaustiveness,
+}
 
-    enum ArgumentType {
-        Exhaustive,
+impl Exhaustive {
+    fn is_conditional(&self) -> bool {
+        matches!(self.kind, Exhaustiveness::Conditional)
     }
-    let mut next_expected: Option<ArgumentType> = None;
-
-    fn handle_next_expected(
-        next_expected: &Option<ArgumentType>,
-        default_value: &mut Option<TokenStream2>,
-        token_stream: TokenStream2,
-    ) {
-        match next_expected {
-            None => panic!("bitenum!: Seen {}, but didn't expect anything. Example of valid syntax: #[bitenum(u3, exhaustive = false)]", token_stream),
-            Some(ArgumentType::Exhaustive) => {
-                *default_value = Some(token_stream);
-            }
-        }
+    fn matches(&self, expected: bool) -> bool {
+        use Exhaustiveness::{False, True};
```

Comment:
remove the "use" here? Can just fully qualify in the line below

Reply:
It's only so that rustfmt doesn't expand the following if/else into several lines. I'll swap back to use the qualified path